### ISSUE TITLE
System.Net.Sockets.APMServer.Tests crashed in CI on Windows

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#4037]
http://dotnet-ci.cloudapp.net/job/dotnet_corefx_windows_debug_prtest/5448/console
```
Discovering: System.Net.Sockets.APMServer.Tests
  Discovering: System.Net.NetworkInformation.Functional.Tests
  Discovered:  System.Net.NetworkInformation.Functional.Tests
  Starting:    System.Net.NetworkInformation.Functional.Tests
  Discovered:  System.Net.Sockets.APMServer.Tests
  Starting:    System.Net.Sockets.APMServer.Tests
  Finished:    System.Net.NetworkInformation.Functional.Tests
  
  === TEST EXECUTION SUMMARY ===
     System.Net.NetworkInformation.Functional.Tests  Total: 19, Errors: 0, Failed: 0, Skipped: 0, Time: 0.547s
  System.Xml.RW.XmlSystemPathResolver.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.Xml.RW.XmlSystemPathResolver.Tests\System.Xml.RW.XmlSystemPathResolver.Tests.dll
  System.Xml.RW.XmlReader.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.Xml.RW.XmlReader.Tests\System.Xml.RW.XmlReader.Tests.dll
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.
  
  Discovering: System.Xml.RW.XmlSystemPathResolver.Tests
  Discovered:  System.Xml.RW.XmlSystemPathResolver.Tests
d:\j\workspace\dotnet_corefx_windows_debug_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00107\lib\tests.targets(129,5): warning MSB3073: The command "CoreRun.exe xunit.console.netcore.exe System.Net.Sockets.APMServer.Tests.dll  -xml testResults.xml -notrait category=nonwindowstests  -notrait category=OuterLoop -notrait category=failing " exited with code -2146232797. [d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Net.Sockets.Legacy\tests\FunctionalTests\System.Net.Sockets.APMServer.Tests.csproj]
d:\j\workspace\dotnet_corefx_windows_debug_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00107\lib\tests.targets(136,5): error : One or more tests failed while running tests from 'System.Net.Sockets.APMServer.Tests' please check log for details! [d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Net.Sockets.Legacy\tests\FunctionalTests\System.Net.Sockets.APMServer.Tests.csproj]
```

**Comment by CIPop** [comment]
@weshaggard @stephentoub @pgavlin I can't find anything in the log outside of what's already noted in the above message. 
>One or more tests failed while running tests from 'System.Net.Sockets.APMServer.Tests' please check log for details! [d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Net.Sockets.Legacy\tests\FunctionalTests\System.Net.Sockets.APMServer.Tests.csproj]

Please reopen if this is failing again or if there is a way to further analyze this particular failure.

I have opened #4152 to track the addition of dump and ETW collection from the CI system.

**Comment by stephentoub** [comment]
Failed again here:
http://dotnet-ci.cloudapp.net/job/dotnet_corefx_windows_debug_prtest/5747/console
```
 Discovering: System.Net.Sockets.APM.Performance.Tests
  Discovered:  System.Net.Sockets.APM.Performance.Tests
  Info:        System.Net.Sockets.APM.Performance.Tests has no tests to run
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.
  
  Discovering: System.Net.Sockets.Tests
  Discovered:  System.Net.Sockets.Tests
  Starting:    System.Net.Sockets.Tests
d:\j\workspace\dotnet_corefx_windows_debug_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00111\lib\tests.targets(129,5): warning MSB3073: The command "CoreRun.exe xunit.console.netcore.exe System.Net.Sockets.APMServer.Tests.dll  -xml testResults.xml -notrait category=nonwindowstests  -notrait category=OuterLoop -notrait category=failing " exited with code -2146232797. [d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Net.Sockets.Legacy\tests\FunctionalTests\System.Net.Sockets.APMServer.Tests.csproj]
d:\j\workspace\dotnet_corefx_windows_debug_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00111\lib\tests.targets(136,5): error : One or more tests failed while running tests from 'System.Net.Sockets.APMServer.Tests' please check log for details! 
```

**Comment by pgavlin** [comment]
The return code implies that this is a failfast (HRESULT is 0x80131623, which is COR_E_FAILFAST). @stephentoub is there any way to get a stack trace on failfast?

**Comment by davidsh** [comment]
@pgavlin It's possible that FAILFAST is happening due to Debug.Assert() being fired.  There was a recent change in the the Debug class to no longer fire FAILFAST.  Instead, it writes out info to stderr.

So, I would suggest trying to repro this again using current code.

**Comment by stephentoub** [comment]
My guess is that this is a Debug.Assert, which has been using Environment.FailFast if not debugger is attached.  I recently changed it to instead throw an exception, but most of the libraries that use System.Diagnostics.Debug haven't been updated to use a newer version of the package that includes my change.

Environment.FailFast does write the message (which would include the stack trace) to the event log.  I'll see if I can grab it off the CI machine, if the log is still there.

**Comment by stephentoub** [comment]
> I'll see if I can grab it off the CI machine, if the log is still there.

Unfortunately the event log didn't have anything for this.

**Comment by CIPop** [comment]
I'm going to tentatively close this as a duplicate of #4204. I've been running the System.Net.Sockets tests (8 instances on a dual-core VM) with that fix for >24hrs with no crashes.

As a precaution, I'll start running the APM tests in parallel on the same VM and in the same configuration.

**Comment by CIPop** [comment]
No issues noted after 21hrs of 8 instances of System.Net.Sockets.APMServer tests running in parallel on the same 2 core VM.
